bpf/lib/encrypt.h

```diff
@@ -205,6 +206,91 @@ do_decrypt(struct __ctx_buff *ctx, __u16 proto)
 #endif /* ENABLE_ROUTING */
 }
 
+static __always_inline int
+ipsec_maybe_redirect_to_encrypt(struct __ctx_buff *ctx, __be16 proto)
+{
+	struct remote_endpoint_info *dst = NULL;
+	struct remote_endpoint_info __maybe_unused *src = NULL;
+	void *data __maybe_unused, *data_end __maybe_unused;
+	struct iphdr __maybe_unused *ip4;
+	__u32 magic __maybe_unused = 0;
+	int ret = 0;
+	__u8 dst_mac = 0;
+
+	if (!eth_is_supported_ethertype(proto))
+		return DROP_UNSUPPORTED_L2;
+
+	switch (proto) {
+# ifdef ENABLE_IPV4
+	case bpf_htons(ETH_P_IP):
+		if (!revalidate_data(ctx, &data, &data_end, &ip4))
+			return DROP_INVALID;
+
+#  if defined(TUNNEL_MODE)
+		/* In tunneling mode IPsec needs to encrypt tunnel traffic,
+		 * so that src sec ID can be transferred.
+		 *
+		 * This also handles IPv6, as IPv6 pkts are encapsulated w/
+		 * IPv4 tunneling.
+		 */
+		if (ctx_is_overlay(ctx)) {
+			ret = set_ipsec_encrypt(ctx, 0, ip4->daddr,
+						get_identity(ctx), false,
+						true);
```

Comment:
Although I can't find a counterexample in my mind: is it a correct assumption that all pod-to-pod traffic will be encapped to vxlan in a vxlan+ipsec cilium?
Because in theory we should lookup ipcache to determine if the original packets are pod2pod, but now we go for this shortcut. Let's make sure we don't overlook any cornor case.

Reply:
when ipsec + egress gateway, to-netdev can see a vxlan skb but we shouldn't do ipsec encryption in this case, am I correct?

Reply:
I think you are correct since EGW utilizes VXLAN in its implementation. We may want to ask more specifically with someone who is intimate with EGW, but that seems right to me. We could be in VXLAN for EGW and I do not believe that traffic which is utilizing VXLAN for EGW purposes is subjected to encryption. We should confirm tho.

Reply:
I see that we have access to the source security ID for tunnel packets, as `bpf_overlay` sets the src ID and a FROM_OVERLAY host mark. But I'm not sure if this helps, since we'd need some idea that the traffic is for EGW purposes.

Reply:
`to-netdev` will see quite a bunch of VXLAN traffic that's not perfectly associated with pod-to-pod traffic (EGW, DSR-Geneve, North/South LB-to-backend, ...). WG took a pragmatic approach to this, and it resulted in [this nice list](https://docs.cilium.io/en/latest/security/network/encryption-wireguard/#which-traffic-is-encrypted) of cases :grin:.

I wouldn't give this too much thought for now - if our encryption path can transparently handle such traffic, it's probably not worth adding all sorts of special "exclude from IPsec" detection for it.

As an example for EGW: on the GW node, this traffic would *still* end up in the `from-overlay` program after decryption, and receive its EGW handling there same as before.

Reply:
I agree with the idea that "if our encryption path can transparently handle such traffic", but I was concerned about unexpected behavior changes: e.g. customers may assume egw won't be affected by cilium upgrade, but in fact it could: previously egw vxlan is plain-text, after upgrade it may be ipsec encrypted.

Reply:
Ack. I think there we are lucky that most (all?) of these "unexpected" cases are associated with `ENABLE_NODEPORT`, which is still a very **new** feature in combination with `ENABLE_IPSEC`.